src/main/scala/org/scanet/optimizers/AdaDelta.scala

```diff
@@ -0,0 +1,31 @@
+package org.scanet.optimizers
+
+import org.scanet.core._
+import org.scanet.syntax._
+
+case class AdaDelta(rate: Float = 1, rho: Float = 0.9f) extends Algorithm {
+
+  def initMeta[X: TensorType](initArg: Tensor[X]): Tensor[Float] = {
+    Tensor.zeros[Float](Shape(2 :: initArg.shape.dims))
+  }
+
+  def delta(grad: Output[Float], meta: Output[Float]): Delta = {
+    // root mean squared
+    def rms(x: Output[Float]): Output[Float] =
+      (x + 1e-7f.const).sqrt
+    // running (decaying) average
+    def avg(prev: Output[Float], curr: Output[Float]): Output[Float] =
+      rho.const * prev + (1 - rho).const * curr.pow(2)
+
+    val prevAvgGrad = meta.slice(0)
+    val prevAvgDelta = meta.slice(1)
```

Comment:
Could you extract the logic when you unpack and pack metadata back into separate methods, it would make an alg cleaner?